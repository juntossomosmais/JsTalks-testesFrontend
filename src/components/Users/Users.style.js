import styled from 'styled-components'

export const ListWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-bottom: 30px;
  width: 450px;

  ul {
    margin: 0;
    padding: 0;
    list-style: none;
    width: 100%;

    li {
      display: flex;
      border-bottom: solid 1px #95a5a6;
      padding: 15px 0;
      justify-content: space-between;

      &:last-of-type {
        border: none;
      }
    }
  }
`

export const ListActions = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-bottom: 30px;
  width: 100%;
`

export const Button = styled.button`
  background: #34495e;
  border-radius: 4px;
  font-size: 12px;
  color: #fff;
  padding: 8px 12px;
  outline: none;
  margin: 0 10px;

  &:last-of-type {
    margin: 0;
  }
`
