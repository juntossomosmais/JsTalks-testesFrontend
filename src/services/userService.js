import axios from "axios";

export const userService = async () => {
  try {
    const response = await axios.get(
      "https://jsonplaceholder.typicode.com/users"
    );
    return response.data;
  } catch (error) {
    return error;
  }
};
