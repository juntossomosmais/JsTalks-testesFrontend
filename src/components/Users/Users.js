import React, { useState, useEffect } from "react";
import { userService } from "../../services/userService";

import * as Styled from "./Users.style";

const Users = () => {
  const [usersState, setUsersState] = useState([]);

  const orderBy = (list, param) => {
    const compare = (a, b) => {
      const propA = a[param].toString().toUpperCase();
      const propB = b[param].toString().toUpperCase();
      let comparison = 0;

      if (propA > propB) comparison = 1;
      if (propA < propB) comparison = -1;

      return comparison;
    };

    const orderedList = list.sort(compare);

    setUsersState([...orderedList]);
  };

  useEffect(() => {
    const fetchUsers = async () => {
      const users = await userService();
      setUsersState(users);
    };
    fetchUsers();
  }, []);

  return (
    <Styled.ListWrapper>
      <Styled.ListActions>
        <Styled.Button
          data-testid="order-by-name"
          onClick={() => orderBy(usersState, "name")}
        >
          Order By: name
        </Styled.Button>
        <Styled.Button
          data-testid="order-by-email"
          onClick={() => orderBy(usersState, "email")}
        >
          Order By: email
        </Styled.Button>
      </Styled.ListActions>
      <ul data-testid="user-list">
        {usersState.map((user) => (
          <li key={user.id}>
            {user.name} <small>{user.email}</small>
          </li>
        ))}
      </ul>
    </Styled.ListWrapper>
  );
};

export default Users;
