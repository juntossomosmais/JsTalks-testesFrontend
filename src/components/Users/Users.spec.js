import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";

import * as service from "../../services/userService";
import Users from "./Users";

describe("<Users />", () => {
  beforeEach(() => {
    const mockDataFromApi = {
      data: [
        {
          id: 1,
          name: "Leanne Graham",
          username: "Bret",
          email: "atest@april.biz",
          address: {
            street: "Kulas Light",
            suite: "Apt. 556",
            city: "Gwenborough",
            zipcode: "92998-3874",
            geo: { lat: "-37.3159", lng: "81.1496" },
          },
          phone: "1-770-736-8031 x56442",
          website: "hildegard.org",
          company: {
            name: "Romaguera-Crona",
            catchPhrase: "Multi-layered client-server neural-net",
            bs: "harness real-time e-markets",
          },
        },
        {
          id: 2,
          name: "Ervin Howell",
          username: "Antonette",
          email: "ztest@melissa.tv",
          address: {
            street: "Victor Plains",
            suite: "Suite 879",
            city: "Wisokyburgh",
            zipcode: "90566-7771",
            geo: { lat: "-43.9509", lng: "-34.4618" },
          },
          phone: "010-692-6593 x09125",
          website: "anastasia.net",
          company: {
            name: "Deckow-Crist",
            catchPhrase: "Proactive didactic contingency",
            bs: "synergize scalable supply-chains",
          },
        },
      ],
    };
    jest
      .spyOn(service, "userService")
      .mockImplementationOnce(() => Promise.resolve(mockDataFromApi.data));
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("should return the list of users", async () => {
    // when
    const { getByTestId, getByText, rerender } = render(<Users />);
    await waitFor(() => {});
    rerender(<Users />);

    // then
    expect(getByTestId("user-list")).toMatchSnapshot("user list");
    expect(getByTestId("user-list")).toBeInTheDocument();
    expect(getByText("Leanne Graham")).toBeInTheDocument();
    expect(getByText("Ervin Howell")).toBeInTheDocument();
  });
  it("should order items by name", async () => {
    // when
    const { getByTestId, getByText, rerender, queryAllByTestId } = render(
      <Users />
    );
    await waitFor(() => {});
    const orderByName = getByTestId("order-by-name");

    await waitFor(() => fireEvent.click(orderByName));
    rerender(<Users />);

    const names = queryAllByTestId(/user-[0-2]/);

    // then
    expect(getByTestId("user-list")).toMatchSnapshot(
      "user list - ordered by name"
    );
    expect(getByTestId("user-list")).toBeInTheDocument();
    expect(names).toHaveLength(2);
    expect(names[0]).toHaveTextContent("Ervin Howell");
    expect(names[1]).toHaveTextContent("Leanne Graham");
    expect(getByText("Leanne Graham")).toBeInTheDocument();
    expect(getByText("Ervin Howell")).toBeInTheDocument();
  });
  it("should order items by email", async () => {
    // when
    const { getByTestId, getByText, rerender, queryAllByTestId } = render(
      <Users />
    );
    await waitFor(() => {});
    const orderByName = getByTestId("order-by-email");

    await waitFor(() => fireEvent.click(orderByName));
    rerender(<Users />);

    const names = queryAllByTestId(/user-[0-2]/);

    // then
    expect(getByTestId("user-list")).toMatchSnapshot(
      "user list - ordered by email"
    );
    expect(names).toHaveLength(2);
    expect(names[0]).toHaveTextContent("atest@april.biz");
    expect(names[1]).toHaveTextContent("ztest@melissa.tv");
    expect(getByText("Leanne Graham")).toBeInTheDocument();
    expect(getByText("Ervin Howell")).toBeInTheDocument();
  });
});
