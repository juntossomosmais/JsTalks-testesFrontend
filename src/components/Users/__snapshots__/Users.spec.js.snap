// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users /> should order items by email: user list - ordered by email 1`] = `
<ul
  data-testid="user-list"
>
  <li
    data-testid="user-0"
  >
    Leanne Graham
     
    <small>
      atest@april.biz
    </small>
  </li>
  <li
    data-testid="user-1"
  >
    Ervin Howell
     
    <small>
      ztest@melissa.tv
    </small>
  </li>
</ul>
`;

exports[`<Users /> should order items by name: user list - ordered by name 1`] = `
<ul
  data-testid="user-list"
>
  <li
    data-testid="user-0"
  >
    Ervin Howell
     
    <small>
      ztest@melissa.tv
    </small>
  </li>
  <li
    data-testid="user-1"
  >
    Leanne Graham
     
    <small>
      atest@april.biz
    </small>
  </li>
</ul>
`;

exports[`<Users /> should return the list of users: user list 1`] = `
<ul
  data-testid="user-list"
>
  <li
    data-testid="user-0"
  >
    Leanne Graham
     
    <small>
      atest@april.biz
    </small>
  </li>
  <li
    data-testid="user-1"
  >
    Ervin Howell
     
    <small>
      ztest@melissa.tv
    </small>
  </li>
</ul>
`;
